File: frontend/lib/data/models/announcement.dart
enum AnnouncementAudience {
  all('all'),
  category('category'),
  user('user');

  final String apiValue;
  const AnnouncementAudience(this.apiValue);

  static AnnouncementAudience fromApi(String? v) => switch (v) {
        'category' => AnnouncementAudience.category,
        'user' => AnnouncementAudience.user,
        _ => AnnouncementAudience.all,
      };
}

class AdminAnnouncement {
  final String id;
  final String title;
  final String body;
  final AnnouncementAudience audience;
  final String? audienceCategory;
  final String? audienceUserId;
  final int recipients;
  final int pushed;
  final String sentBy;
  final DateTime createdAt;

  const AdminAnnouncement({
    required this.id,
    required this.title,
    required this.body,
    required this.audience,
    this.audienceCategory,
    this.audienceUserId,
    required this.recipients,
    required this.pushed,
    required this.sentBy,
    required this.createdAt,
  });

  factory AdminAnnouncement.fromJson(Map<String, dynamic> j) {
    final aud = (j['audience'] as Map?)?.cast<String, dynamic>() ?? const {};
    return AdminAnnouncement(
      id: j['id'] ?? '',
      title: j['title'] ?? '',
      body: j['body'] ?? '',
      audience: AnnouncementAudience.fromApi(aud['type']),
      audienceCategory: aud['category'] as String?,
      audienceUserId: aud['userId'] as String?,
      recipients: (j['recipients'] ?? 0) as int,
      pushed: (j['pushed'] ?? 0) as int,
      sentBy: j['sentBy'] ?? '',
      createdAt: DateTime.tryParse(j['createdAt'] ?? '') ?? DateTime.now(),
    );
  }
}

File: backend/src/models/Announcement.js
/*
- Announcement model — admin broadcasts (mass messaging).
- One document per broadcast: the source message plus how it was targeted
  and how many recipients it reached. Recipient copies live as per-user
  Notification documents (type 'announcement') so the existing inbox and
  push pipeline stay the single fan-out path.
 */
const mongoose = require('mongoose');
const { CATEGORIES } = require('../config/constants');

const announcementSchema = new mongoose.Schema(
  {
    title: { type: String, required: true, trim: true, maxlength: 120 },
    body: { type: String, required: true, trim: true, maxlength: 2000 },
    audience: {
      type: { type: String, enum: ['all', 'category', 'user'], default: 'all' },
      category: { type: String, enum: CATEGORIES, default: null },
      userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
      phone: { type: String, default: null },
    },
    sentBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    recipients: { type: Number, default: 0 },
    pushed: { type: Number, default: 0 },
  },
  { timestamps: true }
);

announcementSchema.index({ createdAt: -1 });

announcementSchema.methods.toJSON = function () {
  return {
    id: this._id,
    title: this.title,
    body: this.body,
    audience: this.audience,
    sentBy: this.sentBy,
    recipients: this.recipients,
    pushed: this.pushed,
    createdAt: this.createdAt,
  };
};

module.exports = mongoose.model('Announcement', announcementSchema);
